### PR TITLE
Indicate non-profiled mutations for samples in Patient View

### PR DESCRIPTION
**Pull request by pvannierop**
Note: this branch builds on top of branch `patientview_genepanel_api_integration` and the PR is to this branch for the reason of reviewing the code. This PR is therefore in _draft_ mode.

# What? Why?
As defined in RFC45 information on gene panels will be integrated in the PatientView component. Central problem with gene panels in patient view is the ambiguous interpretation of missing mutations in the mutation and copynumber tables. Mutations in samples are here represented as numbered sample icons in respective rows of each table. Absense of a numbered sample icon may be interpreted as absence of a mutation. When gene panels have been used the mutation might not be profiled and the user might incorrectly conclude that the patient sample does not have the mutation.

#  Fix
This PR will improve the interpretation of missing mutations and CNA's in the tables by showing a '?' icon when the gene was 1) not part of a gene panel and 2) shows no mutations/CNA's. Hovering over the '?' icon will shows a tooltip that explains the '?' icon to the user (see screenshots below).

# Area's of change
- A new _dash_ sample icon was introduced. Implementation of this sample icon required refactoring of existing sample icon code to remove code redundancy.

- Logic to show/not show a sample icon or the not-profiled/_dash_ sample icon was implemented. This makes use of gene panel data retrieved from the backend; this feature is implemented in PR #2685. 
- A new study was added for e2e-localdb tests.
- Obsolete study `minimal_study` for e2e-localdb tests was removed.
- Class `sampleManager` was renamed to `SampleManager` (note the capitalization of the _S_)

# Before
image

# After
image
Screenshot from 2019-09-16 15-35-14

### CODE REVIEW

**Comment by inodb** [comment]
Nice 🎉 ! Some small initial visual feedback: I prefer a `-` for unprofiled, the question mark loooks quite big but maybe this was already decided in the RFC?

File: end-to-end-test/shared/specUtils.js [hunk 1]
```diff
@@ -18,6 +18,11 @@ function waitForCoExpressionTab(timeout) {
     $('//*[@id="coexpressionTabGeneTabs"]').waitForExist(timeout || 20000);
 }
 
+function waitForPatientView(timeout) {
+    $('//*[@id="patientViewPageTabs"]').waitForExist(timeout || 20000);
```
Comment (adamabeshouse) [inline end-to-end-test/shared/specUtils.js]: what is this selector syntax?

**Comment by adamabeshouse** [comment]
looks really cool!

File: end-to-end-test/shared/specUtils.js (hunk 1 above)
Comment (Sjoerd-van-Hagen) [reply end-to-end-test/shared/specUtils.js]: If my memory serves me right, all elements anywhere in the hierarchy, that have an attribute id equal to patientViewPageTabs.

**Comment by Sjoerd-van-Hagen** [comment]
> Nice tada ! Some small initial visual feedback: I prefer a `-` for unprofiled, the question mark loooks quite big but maybe this was already decided in the RFC?

We can change the icon if need be. 

@cBioPortal/product any preference for the icon?

File: end-to-end-test/shared/specUtils.js (hunk 1 above)
Comment (adamabeshouse) [reply end-to-end-test/shared/specUtils.js]: hm.. is it equivalent to `#patientViewPageTabs` ?

Comment (Sjoerd-van-Hagen) [reply end-to-end-test/shared/specUtils.js]: If both XPath and CSS selector syntax can be used, then yes, they should be equivalent.

Comment (adamabeshouse) [reply end-to-end-test/shared/specUtils.js]: maybe we can just use # since its better known and equivalent?

Comment (Sjoerd-van-Hagen) [reply end-to-end-test/shared/specUtils.js]: @pvannierop ^

Should be fine, right?

Comment (pvannierop) [reply end-to-end-test/shared/specUtils.js]: I have updated the selector to state `#patientViewPageTabs`.

Comment (adamabeshouse) [reply end-to-end-test/shared/specUtils.js]: can you push the change?

**Comment by jjgao** [comment]
@Sjoerd-van-Hagen: @schultzn suggested "a faint circle with a question mark as the "not profiled" icon" in the RFC. Maybe try that?

**Comment by jjgao** [comment]
Another idea is to change the background of the circle to gray (and maybe add a strikethrough just like oncoprint?)

**Comment by pvannierop** [comment]
I have implemented both. Just indicate your preference.

### circles with question marks
image

### strike-though sample icons
image

File: end-to-end-test/shared/specUtils.js (hunk 1 above)
Comment (pvannierop) [reply end-to-end-test/shared/specUtils.js]: Its pushed AFAIK.

**Comment by Sjoerd-van-Hagen** [comment]
@cBioPortal/product preferences for the icons?

I like Ino's suggestion of using -, because it resembles the oncoprint. The ? is also fine. I do not like the 'circles with question marks' and 'strike-though sample icons' as they are less clear.

I do not think it makes a huge difference though, the mouse over really helps.

**Comment by jjgao** [comment]
I don't like the strikethrough one - bad design by me. The circled question looks ok to be. But I agree that a gray dash may be the best. @pvannierop could you make a mockup?

image

**Comment by pvannierop** [comment]
@jjgao @Sjoerd-van-Hagen 
Here the mockup for the simple dashes.
image

**Comment by jjgao** [comment]
This looks good to me. I think I like the 2nd or 3rd.

**Comment by tmazor** [comment]
I agree the dash looks best - I prefer option 3.

**Comment by pvannierop** [comment]
@jjgao @tmazor I have implemented the dash icon for non-profiled samples.
image

File: src/pages/patientView/mutation/PatientViewMutationTable.tsx [hunk 2]
```diff
@@ -15,6 +15,8 @@ import ExonColumnFormatter from "shared/components/mutationTable/column/ExonColu
 
 export interface IPatientViewMutationTableProps extends IMutationTableProps {
     sampleManager:SampleManager | null;
+    sampleToGenePanelId:{[sampleId: string]: string|undefined};
```
Comment (adamabeshouse) [inline src/pages/patientView/mutation/PatientViewMutationTable.tsx]: so undefined indicates either not profiled or whole exome? or just whole exome?

File: src/pages/patientView/mutation/column/TumorColumnFormatter.tsx [hunk 3]
```diff
@@ -1,35 +1,64 @@
...
+        
         if (!sampleManager) {
             return (<span></span>);
         }
+        
+        // Rules for icon:
+        // - when sample->gene has mutation (present in _mutatedSamples_) show the `sample` icon 
+        // - when sample->gene has no mutation (absent from _mutatedSamples_) and was profiled, show `no mutation` icon
+        // - when sample->gene has no mutation (absent from _mutatedSamples_) and was not profiled, show `not profiled` icon
+        const samples =  sampleManager.samples;
```
Comment (adamabeshouse) [inline src/pages/patientView/mutation/column/TumorColumnFormatter.tsx]: helpful comments

File: src/pages/patientView/mutation/column/TumorColumnFormatter.tsx [hunk 4]
```diff
@@ -67,6 +96,23 @@ export default class TumorColumnFormatter {
...
 
+    public static getProfiledSamplesForGene(entrezGeneId:number, samples:ClinicalDataBySampleId[], sampleToGenePanelId:{[sampleId: string]: string|undefined}, genePanelIdToEntrezGeneIds:{[genePanelId: string]: number[]}) {
+        // For a given gene indicate whether it was profiled in a particular sample
+        return samples.reduce((map, nextSample, currentIndex:number) => {
+            const genePanelId = sampleToGenePanelId[nextSample.id];
+            
+            const wholeGenome = noGenePanelUsed(genePanelId);
+            const isInGenePanel = !wholeGenome && genePanelId && genePanelId in genePanelIdToEntrezGeneIds && genePanelIdToEntrezGeneIds[genePanelId].includes(entrezGeneId);
+
+            if (wholeGenome || isInGenePanel) {
```
Comment (adamabeshouse) [inline src/pages/patientView/mutation/column/TumorColumnFormatter.tsx]: map[nextSample.id] = wholeGenome || isInGenePanel;

also, would be easier to read if renamed `map` to `sampleIsProfiled`

File: src/shared/lib/StoreUtils.spec.ts [hunk 5]
```diff
@@ -498,4 +499,16 @@ describe('StoreUtils', () => {
                 "fetchStudy should be called with the correct study id (study4)");
         });
     });
+
+    describe('wholeGenomeProfiled()', () => {
```
Comment (adamabeshouse) [inline src/shared/lib/StoreUtils.spec.ts]: `describe('noGenePanelUsed` ?

File: src/shared/lib/StoreUtils.spec.ts [hunk 6]
```diff
@@ -498,4 +499,16 @@ describe('StoreUtils', () => {
...
+        it('handles undefined values', () => {
+            assert.isTrue(noGenePanelUsed(undefined));
+        });
+        it('handles "WES"-tag values', () => {
+            assert.isTrue(noGenePanelUsed('WES'));
+        });
+        it('handles "WGS"-tag values', () => {
+            assert.isTrue(noGenePanelUsed('WGS'));
+        });
```
Comment (adamabeshouse) [inline src/shared/lib/StoreUtils.spec.ts]: can you add a test where it returns false?

File: src/pages/patientView/mutation/PatientViewMutationTable.tsx (hunk 2 above)
Comment (pvannierop) [reply src/pages/patientView/mutation/PatientViewMutationTable.tsx]: An `undefined` value for genePanelId means that the sample  was analyzed with a whole genome/exome approach.

File: src/pages/patientView/mutation/column/TumorColumnFormatter.tsx (hunk 3 above)
Comment (pvannierop) [reply src/pages/patientView/mutation/column/TumorColumnFormatter.tsx]: Thnx :)

File: src/pages/patientView/mutation/column/TumorColumnFormatter.tsx (hunk 4 above)
Comment (pvannierop) [reply src/pages/patientView/mutation/column/TumorColumnFormatter.tsx]: > map[nextSample.id] = wholeGenome || isInGenePanel;

Sorry for making this more compicated that needs be, but this change has been applied in a PR builds on top of this PR. So, please disregard this need for code restucturing here.

>also, would be easier to read if renamed map to sampleIsProfiled

Name was updated to `sampleIsProfiled`.

File: src/shared/lib/StoreUtils.spec.ts (hunk 5 above)
Comment (pvannierop) [reply src/shared/lib/StoreUtils.spec.ts]: Ha ha good catch; unit test name was not updated after rename of method name.

File: src/shared/lib/StoreUtils.spec.ts (hunk 6 above)
Comment (pvannierop) [reply src/shared/lib/StoreUtils.spec.ts]: Added!

File: src/pages/patientView/mutation/column/TumorColumnFormatter.tsx (hunk 4 above)
Comment (pvannierop) [reply src/pages/patientView/mutation/column/TumorColumnFormatter.tsx]: I have updated the boolean logic to:
> sampleIsProfiled[nextSampleId] = wholeGenome || isInGenePanel;

**Comment by Sjoerd-van-Hagen** [comment]
@alisman @inodb can we get this merged?

**Comment by alisman** [comment]
@Sjoerd-van-Hagen we agreed yesterday that @pvannierop can merge these when tests pass.  i believe that there is a small data migration which needs to take place and that this is probably and example of a PR that needs to go through rc.  @pvannierop can you comment based our discussion yesterday.

**Comment by alisman** [comment]
@pvannierop @Sjoerd-van-Hagen i cherry picked Pim's backend commits to master.  That's now merged and deployed to master.cbioportal.org.   As Pim predicted, the frontend tests now pass.  The only thing standing in the way is a strange api sync issue.  Working on that now and then I can deploy.

**Comment by alisman** [comment]
@pvannierop @Sjoerd-van-Hagen will create a release and merge/deploy tomorrow.

**Comment by Sjoerd-van-Hagen** [comment]
@alisman Great!

**Pull request merged**